docs/proposals/04-etcd-member-custom-resource.md

```diff
@@ -0,0 +1,545 @@
+
+- Since each backup-sidecar only has knowledge of the health of its own etcd, it can only determine whether its own etcd can be defragmented or not, based on etcd-member health. Trying to defragment a different healthy etcd-member while another etcd-member is unhealthy would lead to a transient quorum loss.
+- Each backup-sidecar is only a `sidecar` to its own etcd-member, and by good design principles, it must not be performing any cluster-wide maintenance operations, and this responsibility should remain with the etcd cluster operator.
+
+Additionally, defragmentation of an etcd DB becomes inevitable if the DB size exceeds the specified DB space quota, since the etcd DB then becomes read-only, ie no write operations on the etcd would be possible unless the etcd DB is defragmented and storage space is freed up. In order to automate this, druid will now need information about the etcd DB size from each member, specifically the leading etcd-member, so that a cluster-wide defragmentation can be triggered if the DB size reaches a certain threshold, as already described by [this issue](https://github.com/gardener/etcd-backup-restore/issues/556).
+
+The authors propose to enhance each etcd-member to regularly publish information about the `dbSize` and ` dbSizeInUse` so that druid may trigger defragmentation for the etcd cluster. 
+
+```yaml
+dbSize: <db-size> # e.g 6Gi
+dbSizeInUse: <db-size-in-use> # e.g 3.5Gi
+```
+
+Difference between `dbSize` and `dbSizeInUse` gives a clear indication of how much storage space would be freed up if a defragmentation is performed. If the difference is not significant (based on a configurable threshold provided to druid), then no defragmentation should be performed. This will ensure that druid does not perform frequent defragmentations that do not yield much benefit. Effectively it is to maximise the benefit of defragmentation since this operations involves transient downtime for each etcd-member. 
+
+### Monitoring Defragmentations
+
+As discussed in the previous section, every etcd-member is defragmented periodically, and can also be defragmented based on the DB size reaching a certain threshold. It is beneficial for druid to have knowledge of this data from each etcd-member for the following reasons:
+
+* [**Diagnostics**] Druid can aggregate data about every member's defragmentations and calculate the frequency of defragmentations on an etcd cluster. Combined with details about the triggers for these defragmentations, druid can derive insights and raise alerts about the cause for frequent defragmentations, such as possible DB spikes.
+
+* [**Operational**] Derive status of defragmentation at etcd cluster level. In case of partial failures for a subset of etcd-members druid can potentially re-trigger defragmentation only for those etcd-members.
+
+The authors propose to capture this information as part of `defragmentations` section in the `EtcdMember` resource.
+
+```yaml
+defragmentations:
```

Comment:
How long will this information be retained and will there be some garbage collection for older defragmentation runs? I'm wondering, this "historical data" is usually more something that is exposed by metrics and scraped by Prometheus.

Reply:
We will only maintain last X run data. As and when newer data arrives, oldest entry will get removed. `X` is not yet been decided and will be decided during implementation.

Reply:
It's understandable to save historic data in the status if it's read again by Druid to determine if something goes south. However for operators and diagnosis it would be better to expose metrics (in addition).

Reply:
Ok we will discuss this point again.

Reply:
As per the proposal we capture:
```yaml
 - startTime: <start time of defragmentation>
    endTime: <end time of defragmentation>
    reason: 
    message:
    initialDBSize: <size of etcd DB prior to defragmentation>
    finalDBSize: <size of etcd DB post defragmentation>
```
Using metrics it would not be very clean to capture all of the above information using a single metric. So you will actually need 2 metrics - one for `initialDBSize` and another for `finalDBSize` and then you need to correlate the two metrics as well. Also message/reason can only be captured via labels (ok for reason as it will be a code, but message will be a description which is not really nice for labels as indexing is done for all defined labels on metrics)

Do you have an alternative way to capture this information via metrics?

Reply:
You are right, you can't map this concept to metrics 1:1. Instead we'd need different metrics, e.g. DB size (already there), maybe a defragmentation failure count and similar. Details for a particular error can still be extracted from logs.

Reply:
Another reason for not pushing these as metrics is because we want druid to take certain actions (mainly remediatory) based on this information, and we'd need to make druid prometheus-aware, which is not a very nice thing to do.

Reply:
> Details for a particular error can still be extracted from logs.

That doesn't seem like a very clean approach, given that druid will then need to fetch etcd pod logs, parse them and extract error information and the exact error codes too. Not to mention that if the etcd pod gets restarted, the logs are lost to druid, unless it then relies on loki for the logs, which means druid will then have to be loki-aware too.

Reply:
I think this is a misunderstanding. Druid should never parse logs or metrics in order to perform some action, see my [previous comment](https://github.com/gardener/etcd-druid/pull/658#discussion_r1298071505):

> It's understandable to save historic data in the status if it's read again by Druid to determine if something goes south. However for operators and diagnosis it would be better to expose metrics (in addition) - same for snapshot information.

Reply:
We can surely look at enhancing metrics for etcd clusters and we will do this during the implementation. I think we will have a much better idea on whether it makes sense to only capture the last defragmentation in the status of an etcd-member vs having an array of X number of entries as proposed currently. This will depend upon what druid would need in the end to take an action.

Reply:
@timuthy we internally discussed and decided that we will only capture last defragmentation and last restoration as part of the `EtcdMember` spec. We have made the changes to these sections. PTAL.